src/osp/Active/drawing.h

```diff
@@ -24,67 +24,202 @@
+    // Non-owning user data passed to draw function, such as the shader
+    void *m_data;
+
+}; // struct EntityToDraw
 
 /**
- * Represents an entity which requests the scene to be drawn to m_target
+ * @brief Tracks a set of entities and their assigned drawing functions
+ *
+ * Each new entity added (in ACtxRenderGroups::m_added) is passed to one of
+ * every RenderGroup's DrawAssigner_t depending on its material type, which is
+ * responsible for setting its draw function and shader.
+ *
+ * DrawAssigners allow each RenderGroup to assign their own shaders for each
+ * kind of material. RenderGroups are intended to be associated with certain
+ * rendering techniques like Forward, Deferred, and Shadow mapping.
+ *
+ * This also works with game-specific modes like Thermal Imaging.
+ */
+struct RenderGroup
+{
+    using Storage_t = entt::storage_traits<ActiveEnt, EntityToDraw>::storage_type;
+    using ArrayView_t = Corrade::Containers::ArrayView<ActiveEnt>;
+
+    /**
+     * @brief Assigns drawing functions and user data (like shaders) to entities
+     *        of a certain material type for each RenderGroup
+     *
+     * @param ActiveScene& [ref] Scene the entities are part of
+     * @param Storage_t&   [ref] Storage to add accepted entities and data to
+     * @param ArrayView_t  [in] Entities to consider
+     */
+    using DrawAssigner_t = std::function<
+            void(ActiveScene&, Storage_t&, ArrayView_t) >;
+
+    /**
+     * @brief Set the DrawAssigner function for a certain material type
+     */
+    template<typename MATERIAL_T>
+    void set_assigner(DrawAssigner_t assigner)
+    {
+        material_id_t id = material_family_t::type<MATERIAL_T>;
+        m_assigners.resize(std::max(m_assigners.size(), size_t(id + 1)));
+        m_assigners[id] = std::move(assigner);
+    }
+
+    /**
+     * @return Iterable view for stored entities
+     */
+    decltype(auto) view()
+    {
+        return entt::basic_view{m_entities};
+    }
+
+    // index with material_id_t
+    std::vector<DrawAssigner_t> m_assigners;
+    Storage_t m_entities;
+
+}; // struct RenderGroup
+
+/**
+ * @brief Stores RenderGroups in the scene
+ */
+struct ACtxRenderGroups
+{
+    ACtxRenderGroups() = default;
+    ACtxRenderGroups(ACtxRenderGroups const&) = delete;
+
+    /**
+     * @brief Resize vectors to fit all specified material IDs
+     */
+    template<typename ... MATERIAL_T>
+    void resize_to_fit()
+    {
+        size_t minSize{std::max({material_family_t::type<MATERIAL_T> ...}) + 1};
+
+        m_added.resize(minSize);
+    }
+
+    /**
+     * @brief Adds a single entity to be considered by RenderGroups and
+     *        assigned draw functions
+     *
+     * @tparam MATERIAL_T Material of newly added entity
+     */
+    template<typename MATERIAL_T>
+    void add(ActiveEnt ent)
+    {
+        m_added[material_family_t::type<MATERIAL_T>].push_back(ent);
+    }
+
+    std::unordered_map< std::string, RenderGroup > m_groups;
```

Comment:
consider an std::unordered_set with a custom search predicate and storing the name inside of the RenderGroup

This cuts down on the number of distinct types the compiler has to deal with, and keeps the name of the object local to the object, making it easier on other parts of the code needing to co-locate the data.

Reply:
Also: shared_string is a good choice for write-once, read-many string types.

Reply:
I'll leave this as-is due to the 'flat' nature of how this project is structured. The RenderGroup isn't really responsible for anything outside of storing its own data. It's unlikely for other code to need to store references to render groups for long term.

I remember a fixed version of shared_string being somewhere, but as of now it still doesn't compile.

Reply:
What...? I fixed it like a month ago.

What errors did you get?